### ISSUE TITLE
Allow loading libgdiplus from NuGet packages

### ISSUE THREAD

**Issue by qmfrederik** [dotnet/corefx#24213]
I just kicked off a first test run of our software, migrating from CoreCompat.System.Drawing to System.Drawing.Common.

On Mac, all unit tests which use System.Drawing are currently red because `libgdiplus.dylib` cannot be located.

Although `libgdiplus.dylib` can be acquired via Homebrew, we try to make it a "plug & play" experience and ship `libgdiplus.dylib` as part of our self-contained app. `libgdiplus.dylib` therefore is not preinstalled on our CI servers; we acquire it via the `runtime.osx.10.10-x64.CoreCompat.System.Drawing/1.0.1-beta004` package.
The same holds true for our end user's desktops - it is currently not a requirement that they run `brew install libgdiplus` before launching our self-contained app.

It appears that the `dlopen` call which is used to load `libgdiplus.dylib` does not use same search path as an `[DllImport]` statement - apparently the NuGet packages are not being searched. 

I'm no expert in this area but it seems the values we're looking for are stored in [m_NativeDllSearchDirectories](https://github.com/dotnet/coreclr/search?utf8=%E2%9C%93&q=m_NativeDllSearchDirectories&type=)

I'd be happy to submit a PR that fixes this behavior but I'm currently not entirely sure where we would fetch the search paths that need to be passed to `dlopen`, hence the issue.

On Linux, this isn't that much of an issue for us because we use, say, `dotnet-rpm` to create RPM packages on CentOS and we can express the libgdiplus dependency in the RPM package - so when installing our package, libgdiplus is automatically installed by the package manager.
On Windows, this isn't an issue at all as GDI+ is always there.

**Comment by mellinoe** [comment]
I don't think there is a clean way to support this right now. We will probably need an interface to the runtime that will enumerate the native search paths for us. At the moment, I don't think we can reliably emulate the search paths that the runtime uses without that. We could special case the application's base path, which would allow published applications to locate and load the binary.

@russellhadley

**Comment by qmfrederik** [comment]
@mellinoe What I did to workaround that was to embed a 'lighthouse' library in my NuGet package. The only thing those libraries contain is a function that echos back the path at which the library is located (see [lighthouse.c](https://github.com/quamotion/adk-nuget/blob/master/AdkNuGetGenerator/lighthouse/osx/lighthouse.c))

So I could figure out the path to the dylib I want System.Drawing.Common to use.

As a next step, I would need to let System.Drawing.Common know; for example by manually calling `SafeNativeMethods.Gdip.LoadNativeLibrary` with a path to the `.dylib`.

Is that an API you'd consider exposing?

**Comment by mellinoe** [comment]
Before doing something like that, I'd like to look at the situation holistically and try to come up with a real solution. Right now, it feels like there is no way to portably handle this use case, which is bad for the whole ecosystem, not just System.Drawing.Common.

@russellhadley Perhaps you and I can sync up about this.

**Comment by mellinoe** [comment]
One thing I didn't mention is that, in my "nativelibraryloader" project, I do handle loading native binaries from nuget assets. https://github.com/mellinoe/nativelibraryloader/blob/master/NativeLibraryLoader/PathResolver.cs#L51

Unfortunately, it relies on Microsoft.Extensions.DependencyModel, which is probably a non-starter for a library in corefx. It uses the runtime dependencies file to locate the native asset, but that would not necessarily be possible if you were using mono or some other runtime.

**Comment by qmfrederik** [comment]
As a FYI for anyone who might be seeing a similar issue:

We migrated from CoreCompat.System.Drawing to System.Drawing.Common as we consider System.Drawing.Common to be the more stable library.

Not being able to load `libgdiplus.dylib` from a NuGet package/redistribute it with our app was immediate cause for helpdesk tickets; so here's what we did instead:

- `runtime.osx.10.10-x64.CoreCompat.System.Drawing` now includes a `libgdiplus-lighthouse.dylib` dylib which exposes a single function, `get_moduleFileName`, which echos back the path of `libgdiplus-lighthouse.dylib`. You can invoke it like this:
```csharp
[DllImport("libgdiplus-lighthouse.dylib", EntryPoint = "get_moduleFileName", CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Ansi)]
public static extern int GetGdiPlusPath(StringBuilder moduleFileName, int moduleFileNameLength);
```

- From here, you can calculate the path to `libgdiplus.dylib`. A basic implementation would be:
```csharp
public static IntPtr LoadNativeLibrary()
{
    // Emulate the CoreFX behavior
    var handle = dlopen("libgdiplus.dylib", RTLD_NOW);

    // If that didn't work, then try to load libgdiplus from the NuGet package.
    if (handle == IntPtr.Zero)
    {
        StringBuilder builder = new StringBuilder(1024);
        int byteCount = GetGdiPlusPath(builder, builder.Capacity);

        string lighthouseFileName = builder.ToString();
        string gdiPlusDirectory = Path.GetDirectoryName(lighthouseFileName);
        string gdiPlusPath = Path.Combine(gdiPlusDirectory, "libgdiplus.dylib");

        handle = dlopen(gdiPlusPath, RTLD_NOW);
    }

    return handle;
}
```

You can then swizzle (replace) `System.Drawing.SafeNativeMethods+Gdip.LoadNativeLibrary` with your copy of `LoadNativeLibrary`, e.g. by following the procedure described here: http://www.codeproject.com/Articles/37549/CLR-Injection-Runtime-Method-Replacer

This is from the "it works" department - the code is not very elegant and can break whenever the System.Drawing.Common internals change.

Long story short to say that we're unblocked at the moment, but it would be better if System.Drawing.Common provided a built-in way to do this :-).

**Comment by weshaggard** [comment]
@jkotas @RussKeldorph Do one of you guys know the best way to load a native assembly from the application directory that is in the runtime split layout?

**Comment by jkotas** [comment]
`AppContext.GetData("NATIVE_DLL_SEARCH_DIRECTORIES")` will give you the list of native probing directories setup by the host.

**Comment by jkotas** [comment]
I think you meant to cc @russellhadley

**Comment by RussKeldorph** [comment]
> I think you meant to cc @russellhadley

I concur. :)
